cli/addressbook/address.go

```diff
@@ -0,0 +1,107 @@
+package addressbook
+
+import (
+	"fmt"
+
+	"github.com/AlecAivazis/survey/v2"
+	ethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/cobra"
+
+	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/helpers"
+	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/types"
+)
+
+const (
+	// ethAddressCfgKey is an optional key passed in the config files.
+	ethAddressCfgKey = "eth_address"
+)
+
+type addressConfig struct {
+	EthAddress string `mapstructure:"eth_address,omitempty"`
+}
+
+type Address struct {
+	address types.Address
+	cfg     *addressConfig
+}
+
+func newAddress(address types.Address, cfg *addressConfig) (*Address, error) {
+	return &Address{address: address, cfg: cfg}, nil
+}
+
+type AddressFactory struct{}
+
+func (a *AddressFactory) unmarshalConfig(raw map[string]interface{}) (*addressConfig, error) {
+	if raw == nil {
+		// No eth_address defined.
+		return &addressConfig{}, nil
+	}
+
+	var cfg addressConfig
+	if err := mapstructure.Decode(raw, &cfg); err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+func (a *AddressFactory) Load(_ string, rawCfg map[string]interface{}, rawAddr string) (*Address, error) {
+	cfg, err := a.unmarshalConfig(rawCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	addr, _, err := helpers.ResolveEthOrOasisAddress(rawAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	return newAddress(*addr, cfg)
+}
+
+func (a *AddressFactory) Add(cfg map[string]interface{}, address string) (*Address, error) {
+	addr, ethAddr, err := helpers.ResolveEthOrOasisAddress(address)
+	if err != nil {
+		return nil, err
+	}
+
+	ethAddrHex := ""
+	if ethAddr != nil {
+		ethAddrHex = ethAddr.Hex()
+		cfg[ethAddressCfgKey] = ethAddrHex
+	}
+
+	return newAddress(*addr, &addressConfig{
+		EthAddress: ethAddrHex,
+	})
+}
+
+func (a *AddressFactory) DataPrompt() survey.Prompt {
+	return &survey.Multiline{Message: "Address (bech32 or hex-encoded):"}
+}
+
+func (a *AddressFactory) DataValidator() survey.Validator {
+	return func(ans interface{}) error {
+		if addr, _, err := helpers.ResolveEthOrOasisAddress(ans.(string)); addr == nil {
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("unsupported address format")
+		}
+		return nil
+	}
+}
+
+func (a *Address) Address() types.Address {
+	return a.address
+}
+
+func (a *Address) EthAddress() *ethCommon.Address {
+	if a.cfg.EthAddress != "" {
+		_, ethAddr, err := helpers.ResolveEthOrOasisAddress(a.cfg.EthAddress)
```

Comment:
This should be handled during loading from config and then stored in a private field. Also during `Load` it should be checked that the "eth address" actually corresponds to the raw address.